src/memray/_memray/tracking_api.cpp

```diff
@@ -300,6 +310,97 @@ PythonStackTracker::popPythonFrame(PyFrameObject* frame)
     invalidateMostRecentFrameLineNumber();
 }
 
+void
+PythonStackTracker::installGreenletTraceFunctionIfNeeded()
+{
+    if (!s_greenlet_tracking_enabled || d_greenlet_hooks_installed) {
+        return;  // Nothing to do.
+    }
+
+    assert(PyGILState_Check());
+
+    RecursionGuard guard;
+
+    // Borrowed reference
+    PyObject* modules = PySys_GetObject("modules");
+    if (!modules) {
+        return;
+    }
+
+    // Borrowed reference
+    // Look directly at `sys.modules` since we only want to do something if
+    // `greenlet._greenlet` has already been imported.
+    PyObject* _greenlet = PyDict_GetItemString(modules, "greenlet._greenlet");
+    if (!_greenlet) {
+        return;
+    }
+
+    // Borrowed reference
+    PyObject* _memray = PyDict_GetItemString(modules, "memray._memray");
+    if (!_memray) {
+        return;
+    }
+
+    PyObject* ret = PyObject_CallMethod(
+            _greenlet,
+            "settrace",
+            "N",
+            PyObject_GetAttrString(_memray, "greenlet_trace_function"));
+    Py_XDECREF(ret);
+
+    if (!ret) {
+        // This might be hit from PyGILState_Ensure when a new thread state is
+        // created on a C thread, so we can't reasonably raise the exception.
```

Comment:
I mean, we *could* raise an exception because at that time we have the GIL, no? Is just that the exception will be ignored

Reply:
The exception will be ignored, because `PyGILState_Ensure` isn't expected to return with the exception indicator set, and so the caller won't check for it. The caller will go on to make Python C API calls with the exception indicator already set, in violation of the C API's contract.